Misc. support for nrf54h20dk_nrf54h20_cpurad_iron

SebastianBoe opened the pull request:
Misc. support for nrf54h20dk_nrf54h20_cpurad_iron.

See commit messages for details.

nordicjm commented on soc/nordic/nrf54h/cpurad_mgmt.c:
```diff
@@ -0,0 +1,22 @@
+	return ironside_se_cpuconf_boot_radiocore((uint8_t*)&dummy, sizeof(dummy));
+}
+
+SYS_INIT(nrf54h20_cpurad_init, POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE);
```
use soc hook

SebastianBoe replied:
I want the application system call that starts the network/radio CPU to be as consistent as possible between nrf53 and nrf54h20.

On nrf53 it is done like this:

SYS_INIT(nrf53_cpunet_init, POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE);

Shall I modify nrf53 to also use soc hook, or is there some reason you would like it different from 53?

nordicjm replied:
Yes, `SYS_INIT()` is legacy, all should be moved over to soc hooks, can see an example here: https://github.com/zephyrproject-rtos/zephyr/pull/78429/files likewise boards should use board hooks

SebastianBoe replied:
done

jonathannilsen committed:
boards: nrf54h0: iron: enable cpusec_bellboard for cpuapp

Enable the cpusec_bellboard by default for the cpuapp iron board variant
so that IPC toward the secure domain is available by default. Because
IPC is needed to boot the radio core, this fits the standard use case.

Signed-off-by: Jonathan Nilsen <jonathan.nilsen@nordicsemi.no>

SebastianBoe committed:
boards: nordic: add 'iron' board variant for radio

This is needed for next generation Secure Domain firmware.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
soc: nrf54: Port SYS_INIT to use soc_early_init_hook

Port SYS_INIT to use soc_early_init_hook as SYS_INITs are legacy.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
drivers: firmware: nrf_ironside: add IRONside CPUCONF service

Add an IPC service API for booting local domain cores.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
cpuconf: Boot the radiocore from the app in soc_late_init_hook

Boot the radiocore from the app in soc_late_init_hook.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
samples: sysbuild: Add support for nrf54h20dk/nrf54h20/cpuapp/iron

Add support for nrf54h20dk/nrf54h20/cpuapp/iron to the
samples/sysbuild/hello_world sample.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

jonathannilsen commented on include/zephyr/drivers/firmware/nrf_ironside/cpuconf.h:
```diff
@@ -0,0 +1,66 @@
+ * @note cpu_wait is only intended to be enabled for debug purposes
+ * and it is only supported that a debugger resumes the CPU.
+ *
+ * @retval 0 on success or if the CPU has already booted.
+ * @retval -IRONSIDE_CPUCONF_ERROR_WRONG_CPU if cpu is unrecognized
+ * @retval -IRONSIDE_CPUCONF_ERROR_STARTING_CPU_FAILED if starting the CPU failed
+ * @retval -IRONSIDE_CPUCONF_ERROR_MEM_ACCESS_NOT_PERMITTED
+ * if the CPU does not have read access configured for the vector_table address
+ * @retval -IRONSIDE_CPUCONF_ERROR_IRONSIDE_CALL_FAILED The IRONside call failed.
```
Needs a few updates for the latest implementation:

This no longer seems to be used, so can be deleted:
```
* @retval -IRONSIDE_CPUCONF_ERROR_STARTING_CPU_FAILED if starting the CPU failed
```

This is also no longer used from what I can tell, the API returns `buf->status` directly if it's nonzero:
```
* @retval -IRONSIDE_CPUCONF_ERROR_IRONSIDE_CALL_FAILED The IRONside call failed.
```
So we should instead state that the ironside_call_buf status is returned as a positive value if it's nonzero

SebastianBoe replied:
I have a patch for this. But i'd rather apply this after this PR goes in.

SebastianBoe replied:
fixed in https://github.com/zephyrproject-rtos/zephyr/pull/89333

hakonfam commented on boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpurad_iron_0_9_0.yaml:
```diff
@@ -0,0 +1,20 @@
+  - zephyr
+sysbuild: true
+ram: 192
+flash: 256
```
This is 336 kB in `nrf54h20dk_nrf54h20-memory_map_iron.dtsi`

SebastianBoe replied:
I have a patch for this. But i'd rather apply this after this PR goes in.

SebastianBoe replied:
fixed in https://github.com/zephyrproject-rtos/zephyr/pull/89333

hakonfam commented on samples/sysbuild/hello_world/boards/nrf54h20dk_nrf54h20_cpuapp_iron.conf:
```diff
@@ -0,0 +1,4 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
```
nit: 2025?

anangl commented on boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpurad_iron.dts:
```diff
@@ -0,0 +1,24 @@
+#include "nrf54h20dk_nrf54h20-ipc_conf_iron.dtsi"
+#include "nrf54h20dk_nrf54h20-memory_map_iron.dtsi"
+
+/delete-node/ &cpusec_cpuapp_ipc;
```
This is already done in https://github.com/zephyrproject-rtos/zephyr/blob/76e1fc7713a4a3f2b50c497afddec0364a34c10b/boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpurad.dts#L15

SebastianBoe replied:
fixed in https://github.com/zephyrproject-rtos/zephyr/pull/89333

anangl commented on soc/nordic/nrf54h/soc.c:
```diff
@@ -126,22 +126,15 @@ bool z_arm_on_enter_cpu_idle(void)
 
 	trim_hsfll();
 
-	err = dmm_init();
```
After this is removed, also the following line is no longer needed: https://github.com/zephyrproject-rtos/zephyr/blob/12f4ac9dbfbce7ea0fdef2dcd973116ebb78dc13/soc/nordic/nrf54h/soc.c#L24

SebastianBoe replied:
fixed in https://github.com/zephyrproject-rtos/zephyr/pull/89333

anangl commented on soc/nordic/common/dmm.c:
```diff
@@ -295,3 +295,10 @@ int dmm_init(void)
+ * CONFIG_KERNEL_INIT_PRIORITY_OBJECTS. To enable early device use of
+ * dmm, we initialize immediately after kheap.
+ */
+SYS_INIT(dmm_init, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
```
Now `dmm_init()` does not need to be externally accessible, so it can be made static and the changes in `dmm.h` from c84c2fc37dd0d76c5b6735ed72c25de073ea2993 can be reverted.

SebastianBoe replied:
fixed in https://github.com/zephyrproject-rtos/zephyr/pull/89333

anangl commented on include/zephyr/drivers/firmware/nrf_ironside/cpuconf.h:
```diff
@@ -0,0 +1,66 @@
+
+#include <zephyr/drivers/firmware/nrf_ironside/call.h>
+
+#include <nrfx.h>
```
Do you really need nrfx here? For `NRF_PROCESSORID_Type `, `nrf.h` should be sufficient.

anangl commented on drivers/firmware/nrf_ironside/cpuconf.c:
```diff
@@ -0,0 +1,42 @@
+#include <zephyr/sys/util.h>
+#include <zephyr/kernel.h>
+
+#include <nrfx.h>
```
This is redundant here, as `zephyr/drivers/firmware/nrf_ironside/cpuconf.h` includes it.

SebastianBoe replied:
fixed in https://github.com/zephyrproject-rtos/zephyr/pull/89333

anangl commented on soc/nordic/nrf54h/soc.c:
```diff
@@ -43,6 +44,8 @@ sys_snode_t soc_node;
 				      ADDRESS_DOMAIN_Msk |                     \
 				      ADDRESS_BUS_Msk)))
 
+#define dt_nodelabel_cpurad_slot0_partition DT_NODELABEL(cpurad_slot0_partition)
```
Names of macros should use uppercase.

SebastianBoe replied:
fixed in https://github.com/zephyrproject-rtos/zephyr/pull/89333

SebastianBoe commented on include/zephyr/drivers/firmware/nrf_ironside/cpuconf.h:
```diff
@@ -0,0 +1,66 @@
+ * @{
+ */
+
+#define IRONSIDE_CPUCONF_ERROR_STARTING_CPU_FAILED      0x5eb0
+#define IRONSIDE_CPUCONF_ERROR_WRONG_CPU                0x5eb2
+#define IRONSIDE_CPUCONF_ERROR_MEM_ACCESS_NOT_PERMITTED 0x5eb1
+#define IRONSIDE_CPUCONF_ERROR_IRONSIDE_CALL_FAILED     0x5eb3
```
I have a patch for this. But i'd rather apply this after this PR goes in.

karstenkoenig commented on drivers/firmware/nrf_ironside/cpuconf.c:
```diff
@@ -0,0 +1,42 @@
+
+	buf->args[IRONSIDE_CPUCONF_SERVICE_CPU_IDX] = cpu;
+	buf->args[IRONSIDE_CPUCONF_SERVICE_VECTOR_TABLE_IDX] = (uint32_t)vector_table;
+	buf->args[IRONSIDE_CPUCONF_SERVICE_CPU_WAIT_IDX] = cpu_wait;
```
this is using a full uint32_t for a single boolean, maybe this could be a `IRONSIDE_CPUCONF_SERVICE_FLAGS_IDX` and then setting just a single bit for cpu_wait? I can't predict if we will have anything else but that would make it backwards compatible if we ever want to extend the command and is a good basic pattern (collected flags parameter instead of single parameter for each flag)

karstenkoenig replied:
What is the cost if we never need another flag? A single & operation to apply the mask? The way it is now any bit can be set to be 1 in the protocol for this to be evaluated to true, so if we ever want to introduce a new flag we can't do it in this `cpu_wait_idx` field without having to bump to a new version of the cpuconf operation.

SebastianBoe replied:
The cost is an unnecessary abstraction/indirection/container for the reader to understand.

I can update the server to just check the first bit to allow easy introduction of new flags?

karstenkoenig commented on include/zephyr/drivers/firmware/nrf_ironside/cpuconf.h:
```diff
@@ -0,0 +1,66 @@
+	IRONSIDE_CPUCONF_SERVICE_MSG_IDX,
+	IRONSIDE_CPUCONF_SERVICE_MSG_SIZE_IDX,
+	/* The last enum value is reserved for the number of arguments */
+	IRONSIDE_SE_CPUCONF_NUM_ARGS
```
```suggestion
	IRONSIDE_CPUCONF_NUM_ARGS
```

SebastianBoe replied:
fixed in https://github.com/zephyrproject-rtos/zephyr/pull/89333

karstenkoenig commented on include/zephyr/drivers/firmware/nrf_ironside/cpuconf.h:
```diff
@@ -0,0 +1,66 @@
+};
+
+/* IDX 0 is re-used by the error return code and the 'cpu' parameter. */
+#define IRONSIDE_CPUCONF_SERVICE_RETCODE_IDX 0
```
I wonder if we should just define a pattern, like one enum for request indexes and one for response indexes

SebastianBoe replied:
I'd rather not diverge from the update service.

But this sounds like a nice refactoring for when the update service has been merged.

SebastianBoe commented:
These commits are now included in https://github.com/zephyrproject-rtos/zephyr/pull/89333
